File: Server_Code/Packet/LoginNicknameReqPacket.cpp
#include "LoginNicknameReqPacket.h"

LoginNicknameReqPacket::LoginNicknameReqPacket()
    : LoginPacket(EPacketType::LOGIN_NICKNAME_REQ)
{
}

LoginNicknameReqPacket::LoginNicknameReqPacket(const std::string& nickname)
	: LoginPacket(EPacketType::LOGIN_NICKNAME_REQ, nickname)
{
}

std::vector<char> LoginNicknameReqPacket::Serialize() const
{
    BinaryWriter writer;

    writer.WriteHeader(header);
    writer.WriteString(nickname);

    return writer.GetBuffer();
}


void LoginNicknameReqPacket::Deserialize(const char* buffer, size_t bufferSize)
{
    BinaryReader reader(buffer, bufferSize);

    //header = reader.ReadHeader();
    nickname = reader.ReadString();

 
    std::cout << "[Deserialize] Nickname: '" << nickname << "' (" << nickname.length() << " chars)\n";


    if (nickname.empty() || nickname.length() > 16 || nickname.size() > 255)
        throw std::runtime_error("Invalid nickname in LoginNicknameReqPacket");
}

File: Assets/Script/Packet/LoginNicknameReqPacket.cs
using NetworkUtil;
using System;
using System.Text;

public class LoginNicknameReqPacket : LoginPacket
{
    public override EPacketType PacketType => EPacketType.LOGIN_NICKNAME_REQ;

    public LoginNicknameReqPacket() { }

    public LoginNicknameReqPacket(string nickname)
        : base(nickname)
    {
        if (!IsValid())
            throw new ArgumentException("Nickname must be 1~255 bytes in UTF-8.");
    }

    public override void Serialize(BinaryWriter writer)
    {
        writer.WriteString(Nickname);
    }

    public override void Deserialize(BinaryReader reader)
    {
        Nickname = reader.ReadString();
        if (!IsValid())
            throw new ArgumentException("Nickname must be 1~255 bytes in UTF-8.");
    }
}
